basil/HL/bronkhorst_elflow.py

```diff
@@ -16,160 +17,100 @@
-            logger.debug("ELFLOW.set_valve_output() ret error ret=%s" % str(ret))
-            return -1
+        return ret
 
-    def get_valve_output(self):
-        cmd = [4, 114, 0x41, 114, 0x41]
-        self.write(cmd)
-        ret = self.read()
-        if len(ret) != 7:
-            logger.debug("ELFLOW.set_setpoint() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 2 and ret[1] == cmd[1] and ret[2] == cmd[2]:
-            return ((ret[3] << 24) & 0xFF000000) | ((ret[4] << 16) & 0xFF0000) | ((ret[5] << 8) & 0xFF00) | (ret[6] & 0xFF)
-        else:
-            logger.debug("ELFLOW.get_valve_output() ret error ret=%s" % str(ret))
-            return -1
-
-    def set_controller_speed(self, value):
-        value = struct.unpack('<I', struct.pack('<f', value))[0]
-        cmd = [1, 114, 0x40 + 30, (value >> 24) & 0xFF, (value >> 16) & 0xFF, (value >> 8) & 0xFF, value & 0xFF]
-        self.write(cmd)
-        ret = self.read()
-        if len(ret) != 3:
-            logger.debug("ELFLOW.set_controller_speed() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 0 and ret[1] == 0 and ret[2] == 7:
-            return 0
-        else:
-            logger.debug("ELFLOW.set_controller_speed() ret error ret=%s" % str(ret))
-            return -1
-
-    def get_controller_speed(self):
-        cmd = [4, 114, 0x41, 114, 0x40 + 30]
-        self.write(cmd)
+    def get_mode(self):
+        self._intf.write(self.CMDS['get_control_mode'])
         ret = self.read()
-        if len(ret) != 7:
-            logger.debug("ELFLOW.set_setpoint() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 2 and ret[1] == cmd[1] and ret[2] == cmd[2]:
-            return struct.unpack('!f', chr(ret[3]) + chr(ret[4]) + chr(ret[5]) + chr(ret[6]))[0]
-        else:
-            logger.debug("ELFLOW.get_valve_output() ret error ret=%s" % str(ret))
-            return -1
-
-    def get_measure(self):
-        cmd = [4, 1, 0x21, 1, 0x20]
-        self.write(cmd)
+        answer_in_hex = ret[11:]        # read from the 11th digits to translate what mode is on
+        answer = int(answer_in_hex, 16)
+        return answer
+
+    def get_flow(self):
+        """This should give the flow in l/min
+        """
+
+        # first get the max capacity in %
+        self._intf.write(self.CMDS['get_capacity'])
         ret = self.read()
-        if len(ret) != 5:
-            logger.debug("ELFLOW.set_setpoint() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 2 and ret[1] == cmd[1] and ret[2] == cmd[2]:
-            return ((ret[3] << 8) & 0xFF00) | (ret[4] & 0xFF)
-        else:
-            logger.debug("ELFLOW.get_valve_output() ret error ret=%s" % str(ret))
-            return -1
+        answer_in_hex = ret[11:]  # read from the 11th digits to translate what the capacity is
+        cap_100 = struct.unpack('!f', bytes.fromhex(answer_in_hex))[0]
```

Comment:
`cap_100` is not really conclusive. Is this the maximum capacity in percent? Maybe add a comment or find a more descriptive variable name

Reply:
The capacity100 command gives the maximum capacity in the selected capacity unit (This should be L/min). However one can check the unit by using the capacityunit command and by converting the responded hex code into a string. This then should give the unit.